File: sdk/go/wire/wire.go
// Package wire implements the minimal protobuf binary wire primitives used by
// the B-SDK-LANG-EVAL Go golden probe.
//
// Probe scope (route B, hand-written — protoc/protoc-gen-go/buf are not
// installed on this machine): varint scalars, length-delimited values, and
// tag keys only. Fixed32/fixed64 values are skipped on input but never
// emitted; group wire types (3/4) are rejected fail-closed. Zigzag, maps,
// extensions, and service stubs are out of scope. This is a probe, not the
// full SDK protobuf stack.
package wire

import "errors"

// Protobuf wire types.
const (
	TypeVarint  = 0
	TypeFixed64 = 1
	TypeLen     = 2
	TypeSGroup  = 3
	TypeEGroup  = 4
	TypeFixed32 = 5
)

// Errors reported by this package.
var (
	// ErrTruncated reports a buffer that ends in the middle of a field.
	ErrTruncated = errors.New("wire: truncated buffer")
	// ErrOverflow reports a varint longer than the 10-byte 64-bit limit.
	ErrOverflow = errors.New("wire: varint exceeds 10 bytes")
	// ErrBadWire reports a wire type the probe does not support on input
	// (group start/end); decoders fail closed on it.
	ErrBadWire = errors.New("wire: unsupported wire type")
	// ErrFieldZero reports the reserved field number 0.
	ErrFieldZero = errors.New("wire: field number 0")
)

// PutUvarint appends v in minimal protobuf base-128 varint form (1-10 bytes).
// Output is always canonical: no redundant continuation bytes, matching the
// deterministic serialization of protobuf-go.
func PutUvarint(b []byte, v uint64) []byte {
	for v >= 0x80 {
		b = append(b, byte(v)|0x80)
		v >>= 7
	}
	return append(b, byte(v))
}

// Uvarint decodes one base-128 varint from the front of b and returns the
// value together with the number of bytes consumed. Following protobuf
// decoder semantics, non-minimal (padded) encodings are accepted on input;
// truncation and more than 10 bytes are errors.
func Uvarint(b []byte) (uint64, int, error) {
	var v uint64
	for i := 0; i < len(b) && i < 10; i++ {
		c := b[i]
		if i == 9 {
			// The 10th byte contributes exactly one value bit.
			if c > 1 {
				return 0, 0, ErrOverflow
			}
			return v | uint64(c)<<63, 10, nil
		}
		v |= uint64(c&0x7f) << (7 * i)
		if c < 0x80 {
			return v, i + 1, nil
		}
	}
	return 0, 0, ErrTruncated
}

// PutTag appends the varint key packing a field number and wire type.
func PutTag(b []byte, field int, wireType byte) []byte {
	return PutUvarint(b, uint64(field)<<3|uint64(wireType))
}

// PutBytes appends a length-delimited value: varint length prefix + payload.
func PutBytes(b []byte, data []byte) []byte {
	b = PutUvarint(b, uint64(len(data)))
	return append(b, data...)
}

// Bytes reads one length-delimited value from the front of b and returns the
// payload slice (aliasing b) plus the total bytes consumed.
func Bytes(b []byte) ([]byte, int, error) {
	n, header, err := Uvarint(b)
	if err != nil {
		return nil, 0, err
	}
	if uint64(len(b)-header) < n {
		return nil, 0, ErrTruncated
	}
	return b[header : header+int(n)], header + int(n), nil
}

// SkipValue returns the number of bytes occupied by the value of one field
// with the given wire type, starting at the front of b (after its tag).
// Wire types 1 and 5 are skipped so that unknown fields of any future
// additive extension survive a decode/encode cycle; group types are rejected.
func SkipValue(b []byte, wireType byte) (int, error) {
	switch wireType {
	case TypeVarint:
		_, n, err := Uvarint(b)
		return n, err
	case TypeFixed64:
		if len(b) < 8 {
			return 0, ErrTruncated
		}
		return 8, nil
	case TypeLen:
		_, n, err := Bytes(b)
		return n, err
	case TypeFixed32:
		if len(b) < 4 {
			return 0, ErrTruncated
		}
		return 4, nil
	default:
		return 0, ErrBadWire
	}
}

File: sdk/csharp/llmos-sabi-probe/Wire.cs
// Minimal protobuf binary wire primitives for the B-SDK-LANG-EVAL C# golden
// probe (frozen wire v1-beta, ADR-0014).
//
// Probe scope (hand-written; mirrors sdk/go/wire/wire.go): varint scalars,
// length-delimited values, and tag keys only. Fixed32/fixed64 values are
// skipped on input but never emitted; group wire types (3/4) are rejected
// fail-closed. Zigzag, maps, extensions, and service stubs are out of scope.
// This is a probe, not the full SDK protobuf stack.
namespace Llmos.Sabi.Probe;

// Wire-level failure kinds, mirroring the Go probe's sentinel errors.
public enum WireError
{
    Truncated,
    Overflow,
    BadWire,
    FieldZero,
}

public sealed class WireException : Exception
{
    public WireError Kind { get; }

    public WireException(WireError kind, string message) : base(message) => Kind = kind;
}

// Protobuf wire types.
public enum WireType : byte
{
    Varint = 0,
    Fixed64 = 1,
    Len = 2,
    SGroup = 3,
    EGroup = 4,
    Fixed32 = 5,
}

public static class Wire
{
    // Appends v in minimal protobuf base-128 varint form (1-10 bytes).
    // Output is always canonical: no redundant continuation bytes, matching
    // the deterministic serialization the frozen goldens were pinned with.
    public static void PutUvarint(List<byte> b, ulong v)
    {
        while (v >= 0x80)
        {
            b.Add((byte)(v | 0x80));
            v >>= 7;
        }
        b.Add((byte)v);
    }

    // Decodes one base-128 varint at b[pos]. Following protobuf decoder
    // semantics, non-minimal (padded) encodings are accepted on input;
    // truncation and more than 10 bytes are errors. The 10th byte may only
    // carry the final value bit.
    public static (ulong Value, int Consumed) Uvarint(byte[] b, int pos)
    {
        ulong v = 0;
        for (int i = 0; i < 10; i++)
        {
            if (pos + i >= b.Length)
            {
                throw new WireException(WireError.Truncated, "wire: truncated buffer");
            }
            byte c = b[pos + i];
            if (i == 9)
            {
                if (c > 1)
                {
                    throw new WireException(WireError.Overflow, "wire: varint exceeds 10 bytes");
                }
                return (v | (ulong)c << 63, 10);
            }
            v |= (ulong)(c & 0x7f) << (7 * i);
            if (c < 0x80)
            {
                return (v, i + 1);
            }
        }
        throw new WireException(WireError.Truncated, "wire: truncated buffer");
    }

    // Appends the varint key packing a field number and wire type.
    public static void PutTag(List<byte> b, int field, WireType wt) =>
        PutUvarint(b, ((ulong)field << 3) | (ulong)wt);

    // Appends a length-delimited value: varint length prefix + payload.
    public static void PutBytes(List<byte> b, byte[] data)
    {
        PutUvarint(b, (ulong)data.Length);
        b.AddRange(data);
    }

    // Reads one length-delimited value at b[pos]; returns the payload as a
    // fresh copy plus the total bytes consumed.
    public static (byte[] Payload, int Consumed) Bytes(byte[] b, int pos)
    {
        (ulong n, int header) = Uvarint(b, pos);
        if (b.Length - pos - header < (long)n)
        {
            throw new WireException(WireError.Truncated, "wire: truncated buffer");
        }
        var payload = new byte[n];
        Array.Copy(b, pos + header, payload, 0, (long)n);
        return (payload, header + (int)n);
    }

    // Returns the number of bytes occupied by the value of one field with
    // the given wire type, starting at b[pos] (after its tag). Wire types
    // 1 and 5 are skipped so that unknown fields of any future additive
    // extension survive a decode/encode cycle; group types are rejected.
    public static int SkipValue(byte[] b, int pos, WireType wt)
    {
        switch (wt)
        {
            case WireType.Varint:
                return Uvarint(b, pos).Consumed;
            case WireType.Fixed64:
                if (b.Length - pos < 8)
                {
                    throw new WireException(WireError.Truncated, "wire: truncated buffer");
                }
                return 8;
            case WireType.Len:
                (_, int n) = Bytes(b, pos);
                return n;
            case WireType.Fixed32:
                if (b.Length - pos < 4)
                {
                    throw new WireException(WireError.Truncated, "wire: truncated buffer");
                }
                return 4;
            default:
                throw new WireException(WireError.BadWire, "wire: unsupported wire type");
        }
    }
}
